### PR TITLE
Implement autorunning of child loops

### PR DESCRIPTION
**Pull request by wch**
This PR makes it so that child loops are run automatically when a parent loop is run.

When running an event loop, this checks if the event loop has any children, and runs them as well. It also modifies `nextTimestamp()` and `due()` methods so that, when they are called on a registry (AKA event loop), they recurse into children and find the nearest time stamp, or whether any callbacks are due, respectively.

All of this happens in purely C++ code; if we had to call back into R to check for parent-child relationships each time we did this stuff, it would be expensive.

Update: This previously was implemented with external pointers, but it turns out to be much simpler to pass around loop IDs, like we did previously.

### CODE REVIEW

File: R/later.R [hunk 1]
```diff
@@ -47,28 +51,52 @@
...
-
+create_loop <- function(parent = current_loop(), autorun = NULL) {
   id <- .globals$next_id
   .globals$next_id <- id + 1L
-  createCallbackRegistry(id)
+
+  if (!is.null(autorun)) {
+    # This is for backward compatibility, if `create_loop(autorun=FALSE)` is called.
+    parent <- NULL
+    warning("create_loop(autorun=FALSE) is deprecated. Please use create_loop(parent=NULL) from now on.")
```
Comment (jcheng5) [inline R/later.R]: Let's not warn here--if people have CRAN packages calling `create_loop(autorun=FALSE)` let's not make them resubmit just to squelch the warning

File: src/callback_registry.cpp [hunk 2]
```diff
@@ -253,7 +253,28 @@ void testCallbackOrdering() {
...
+  return id;
+}
+
+
+void CallbackRegistry::signal(bool recursive) {
+  Guard guard(mutex);
+  condvar.signal();
+
+  if (recursive && parent != nullptr) {
+    parent->signal();
```
Comment (jcheng5) [inline src/callback_registry.cpp]: `signal(TRUE)`

File: src/callback_registry_table.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    Guard guard(mutex);
+    if (!exists(id)) {
+      return shared_ptr<CallbackRegistry>();
+    }
+    // If the target of the shared_ptr has been deleted, then this is an empty
+    // shared_ptr.
+    return registries[id];
+  }
+
+  bool remove(int id) {
```
Comment (wch) [inline src/callback_registry_table.h]: Add `ASSERT_MAIN_THREAD()`

File: src/callback_registry_table.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    Guard guard(mutex);
```
Comment (jcheng5) [inline src/callback_registry_table.h]: ASSERT_MAIN_THREAD

**Comment by jcheng5** [comment]
1. The multiple CallbackRegistry mutexes can't be made safe unless you can guarantee the locks are always acquired in the same order. One way out of this would be to have a single mutex protecting all of later('s data structures).
2. For the global lock approach, all entry points into later would need to be protected by the global lock; `execLaterNative2` and `execCallbacks` included (though the latter would need to be careful NOT to hold the global lock while executing user-supplied callbacks).

File: src/callback_registry.cpp (hunk 2 above)
Comment (wch) [reply src/callback_registry.cpp]: No longer a need for signaling the parent, since we have a shared lock.

**Comment by wch** [comment]
I've implemented the changes we discussed.

**Comment by wch** [comment]
I've been thinking about this, and I'm not sure that I agree with the proposed semantics. One way of looking at it is that, if you manually create a loop and do weird stuff with it, you're responsible for keeping the reference to the loop until you're done with it. Anyone who is doing stuff like this should be able to grasp the issue, though we will have to document it.

For example, here's the problematic case:

```R
callback <- function() {
  message("hello")
}

f <- function() {
  loop <- create_loop()
  later(callback, 1, loop = loop)
}
f()
gc()
# Wait for 1 second - nothing happens
```

In order to fix it, the user can wrap the callback with a function that captures the loop:

```R
g <- function() {
  loop <- create_loop()
  later(function() callback(), 1, loop = loop)
}
g()
gc()
# Wait for 1 second - prints "hello"
```

Also, in cases like WebSocket and Chromote, if someone `rm`'s the object, they probably want it to stop, and not keep doing stuff.

*****

If you still think that we should keep a loop alive when there are callbacks, here's what I think the logic should look like.

* If the loop has a parent, it should be destroyed when:
  * There are no R references to the loop, AND the loop is empty
* If the loop does not have a parent, it should be destroyed when:
  * There are no R references to the loop

**Comment by jcheng5** [comment]
I agree with those bullet points. The "it might or might not still execute depending on when a gc happens to occur" seems like it will be a source of bugs that is both surprising and easy to miss during testing.

File: src/later.cpp [hunk 5]
```diff
@@ -142,7 +142,7 @@ bool deleteCallbackRegistry(int loop_id) {
     Rf_error("Can't delete current loop.");
   }
 
-  return callbackRegistryTable.remove(loop_id);
+  return callbackRegistryTable.notifyRRefDeleted(loop_id);
```
Comment (jcheng5) [inline src/later.cpp:145]:
```suggestion
  return callbackRegistryTable.remove(loop_id);
```

File: src/callback_registry_table.h [hunk 6]
```diff
@@ -48,41 +68,108 @@ class CallbackRegistryTable {
...
+    Guard guard(&mutex);
+
+    std::map<int, RegistryHandle>::iterator it = registries.begin();
+
+    // Iterate over all registries. Remove under the following conditions:
+    // * There are no more R loop handle references to the registry, AND
+    // * The registry is empty, OR the registry has no parent.
+    // This logic is equivalent to the logic describing the function, just in
+    // a different order.
+    while (it != registries.end()) {
```
Comment (jcheng5) [inline src/callback_registry_table.h:147]: Please add a comment indicating that parents always come earlier than their children in the iteration order, so the case of a parent being pruned leading to children/grandchildren being pruned will work within a single iteration.

**Pull request merged**